gossipsub.go

```diff
@@ -1412,135 +1412,182 @@ func (gs *GossipSubRouter) doSendRPC(rpc *RPC, p peer.ID, q *rpcQueue, urgent bo
-		return append(slice, &elems[0])
-	}
-
-	out := slice
-	if len(out) == 0 {
-		out = append(out, &RPC{RPC: pb.RPC{}})
-		out[0].from = elems[0].from
-	}
+// split splits the given RPC If a sub RPC is too large and can't be split
+// further (e.g. Message data is bigger than the RPC limit), then it will be
+// returned as an oversized RPC. The caller should filter out oversized RPCs.
+func (rpc *RPC) split(limit int) iter.Seq[RPC] {
+	return func(yield func(RPC) bool) {
+		nextRPC := RPC{from: rpc.from}
+
+		{
+			nextRPCSize := 0
+
+			messagesInNextRPC := 0
+			messageSlice := rpc.Publish
+
+			// Merge/Append publish messages. This pattern is optimized compared the
+			// the patterns for other fields because this is the common cause for
+			// splitting a message.
+			for _, msg := range rpc.Publish {
+				// We know the message field number is <15 so this is safe.
+				incrementalSize := pbFieldNumberLT15Size + SizeOfEmbeddedMsg(msg.Size())
+				if nextRPCSize+incrementalSize > limit {
+					// The message doesn't fit. Let's set the messages that did fit
+					// into this RPC, yield it, then make a new one
+					nextRPC.Publish = messageSlice[:messagesInNextRPC]
+					messageSlice = messageSlice[messagesInNextRPC:]
+					if !yield(nextRPC) {
+						return
+					}
 
-	for _, elem := range elems {
-		lastRPC := out[len(out)-1]
+					nextRPC = RPC{from: rpc.from}
+					nextRPCSize = 0
+					messagesInNextRPC = 0
+				}
+				messagesInNextRPC++
+				nextRPCSize += incrementalSize
+			}
 
-		// Merge/Append publish messages
-		// TODO: Never merge messages. The current behavior is the same as the
-		// old behavior. In the future let's not merge messages. Since,
-		// it may increase message latency.
-		for _, msg := range elem.GetPublish() {
-			if lastRPC.Publish = append(lastRPC.Publish, msg); lastRPC.Size() > limit {
-				lastRPC.Publish = lastRPC.Publish[:len(lastRPC.Publish)-1]
-				lastRPC = &RPC{RPC: pb.RPC{}, from: elem.from}
-				lastRPC.Publish = append(lastRPC.Publish, msg)
-				out = append(out, lastRPC)
+			if nextRPCSize > 0 {
+				// yield the message here for simplicity. We aren't optimally
+				// packing this RPC, but we avoid successively calling .Size()
+				// on the messages for the next parts.
+				nextRPC.Publish = messageSlice[:messagesInNextRPC]
+				if !yield(nextRPC) {
+					return
+				}
+				nextRPC = RPC{from: rpc.from}
 			}
 		}
 
 		// Merge/Append Subscriptions
-		for _, sub := range elem.GetSubscriptions() {
-			if lastRPC.Subscriptions = append(lastRPC.Subscriptions, sub); lastRPC.Size() > limit {
-				lastRPC.Subscriptions = lastRPC.Subscriptions[:len(lastRPC.Subscriptions)-1]
-				lastRPC = &RPC{RPC: pb.RPC{}, from: elem.from}
-				lastRPC.Subscriptions = append(lastRPC.Subscriptions, sub)
-				out = append(out, lastRPC)
+		for _, sub := range rpc.Subscriptions {
+			if nextRPC.Subscriptions = append(nextRPC.Subscriptions, sub); nextRPC.Size() > limit {
```

Comment:
`nextRPC.Size()` - this is the same issue with calling Size in a loop over and over again as for Publish but probably does not matter much as (from my understanding) new subscriptions as well as control are relatively rare compare to Publish thing

Reply:
Yes exactly. Notice that we flush the messages to publish above. So here we aren't calling .Size on Publish messages again.

Splitting due to control messages should be extremely rare (happy to be proven wrong), so I opted to do the easier thing and keep the old code.

Thinking about this just now, we could do a fast path check before these loops that check if the original RPC without the publish messages is small enough to fit and can be yielded.